src/Stream.php

```diff
@@ -860,6 +860,57 @@ public function partialIntersectionCoerciveWith(int $minIntersectionCount, itera
         return $this;
     }
 
+    // STREAM DEBUG OPERATIONS
+
+    /**
+     * Peek at each element in between other Stream operations
+     * to do some action without modifying the stream.
+     *
+     * Useful for debugging purposes.
+     *
+     * @param callable $callback
+     *
+     * @return $this
+     */
+    public function peek(callable $callback): self
+    {
+        [$this->iterable, $buffer] = Transform::tee($this->iterable, 2);
+        $callback(Stream::of($buffer));
```

Comment:
I think it is more intuitive and easy to work with if the peek operation applies the callback function on each individual item of the stream. It will also work similarly to all the other functions that take callbacks. You can think of `peek` kind of like `map` but it does not cause the side effect of overwriting the value with the result of the function.

For example, I imagine using it like this:
```php
Stream::of(['some', 'input', 'iterable'])
    ->map('strtoupper')
    ->map('sha1')
    ->peek(fn ($x) => print(\strlen($x) . \PHP_EOL))
    ->callForEach(fn ($x) => $doSomeSideEffect($x));
```

In this case, I would expect to see it output:
```php
40
40
40
```

BTW, I like that you were able to use `tee` here to implement this. Nice.

Reply:
Hi @markrogoyski, do you think it is better solution?

When i debug I want to set a breakpoint and play with intermediate result: convert it to array, call some function with it and check its return etc.

And when we get every single element in callback function it is harder to do this checks.

Reply:
There are numerous functions that take a callable and execute the callable on each element of the stream:

- `filter`
- `map`
- `flatMap`
- `callForEach`
- `reindex`
- etc.

I think the intuitive expectation is for non-reduce/summary functions to work on each element of the stream. Prior art of `peek` in other languages also works this way.

You bring up an interesting point that there could be other ways to work with it, so for this instance, I propose two different methods:

- `peek`: Executes callable on each item in the stream
- `peekStream`: Executes callable on the Stream itself

This way `peek` continues the pattern of the callable working on each item, and `peekStream` specifically calls out that it works on the Stream in the name.

Reply:
Yeah, `peekStream()` is a good idea. I'll implement it.